CommunityPromotionScreen/Src/NewPromotionScreenbyDefault/Classes/NPSBDP_UIArmory_PromotionHero.uc

```diff
@@ -601,52 +601,80 @@ function bool CanPurchaseAbilityEx(int Rank, int Branch, name AbilityName, out s
 function int GetAbilityPointCost(int Rank, int Branch)
 {
 	local XComGameState_Unit UnitState;
+	local X2SoldierClassTemplate ClassTemplate;
 	local array<SoldierClassAbilityType> AbilityTree;
 	local bool bPowerfulAbility;
+	local bool bFinalRankAbility;
+	local bool bXCOMRowAbility;
+	local bool bAsResistanceHero;
 	local int AbilityRanks; //Rank is 0 indexed but AbilityRanks is not. This means a >= comparison requies no further adjustments
 	local int AbilityCost;
+	local XComLWTuple Tuple; // Tuple for Issue #10
 
 	UnitState = GetUnit();
+	ClassTemplate = UnitState.GetSoldierClassTemplate();
 	AbilityTree = UnitState.GetRankAbilities(Rank);	
+	AbilityRanks = GetAbilitiesPerRank(UnitState);
+	bPowerfulAbility = class'X2StrategyGameRulesetDataStructures'.default.PowerfulAbilities.Find(AbilityTree[Branch].AbilityName) != INDEX_NONE;
+	bFinalRankAbility = Rank >= ClassTemplate.GetMaxConfiguredRank() - 1;
+	bXCOMRowAbility = Branch >= AbilityRanks; // This will be "false" for 4th row of abilities for base WOTC Faction Heroes, since they have a randomized deck of perks, not a "true" XCOM row.
+	bAsResistanceHero = UnitState.IsResistanceHero() || AbilityRanks == 0;
 
 	if (GetCustomAbilityCost(UnitState, AbilityTree[Branch].AbilityName, AbilityCost))
 	{
-		return AbilityCost;
+		// Do nothing, AbilityCost will already hold the config value.
 	}
-
-	AbilityRanks = GetAbilitiesPerRank(UnitState);
-	if (!UnitState.IsResistanceHero() && AbilityRanks != 0)
+	else if (!bAsResistanceHero && !UnitState.HasPurchasedPerkAtRank(Rank) && !bXCOMRowAbility)
 	{
-		if (!UnitState.HasPurchasedPerkAtRank(Rank) && Branch < AbilityRanks)
-		{
-			// If this is a base game soldier with a promotion available, ability costs nothing since it would be their
-			// free promotion ability if they "bought" it through the Armory
-			return 0;
-		}
+		// If this is a base game soldier with a promotion available, ability costs nothing.
+		// Do nothing, AbilityCost will already hold 0.
 	}
-
-	bPowerfulAbility = (class'X2StrategyGameRulesetDataStructures'.default.PowerfulAbilities.Find(AbilityTree[Branch].AbilityName) != INDEX_NONE);
-	// Use "poweful ability" cost in following situations:	
-
-	// 1. This is an "AWC" (XCOM row) ability:
-	if (bPowerfulAbility && Branch >= AbilityRanks)
+	else if (bPowerfulAbility && bXCOMRowAbility || bAsResistanceHero && (bPowerfulAbility || bFinalRankAbility))
```

Comment:
I have no idea if this is correct. I would prefer not to rely on contributors knowing which order logical operators are processed in. More parentheses?

Reply:
I also think `bFinalRankAbility` is wrong. The base game grants a base AP amount of "powerful ability AP cost" to Colonel and any rank above it. Whereas this seems to only apply the powerful-ability cost on the last rank, which might be Brigadier or equivalent. Not entirely sure because I'm having trouble decoding what this logical expression means.

Reply:
Well this is technically a change in the base game behavior, but the base game also doesn't have brigadier rank, so it's kinda muddy. I did post a question regarding desired behavior for the brigadier rank in discord earlier, so this was one of the things I wanted to get your input on. 

I was hoping the self-commenting nature of the code was easy to enough to follow, so I didn't leave verbose comments.

The general idea is that the ability should use the increased point cost in the following case:

1. If it's a "powerful ability" in the non-class ability row.
2. For Resistance Heroes, if this is a "powerful ability" anywhere in their skill tree, though normally they would be only in the 4th "XCOM" row.
3. For Resistance Heroes, if this is a "final rank" / Colonel+ ability.

I opted to go for the "Final Rank" ability in this case since it will work for any number ranks, without any assumptions beyond that there are more than one rank. 

I could change it into `bColonelPlus` if this is the behavior we want for Brigadiers, though personally I'd be less comfortable with it.

Reply:
And yeah I don't blame you, the original code is a nightmare to navigate, though I'm reasonably confident in my implementation relaying its functionality, though I don't mind trying to simplify it even further.

Reply:
> I opted to go for the "Final Rank" ability in this case since it will work for any number ranks, without any assumptions beyond that there are more than one rank.

So the issue with this is that ranking up to Colonel will grant 25 AP on top of the base promotion award (see `XCGS_Unit.GetResistanceHeroAPAmount()`, but the promotion screen does not reflect this in that rank's AP costs (unless the abilities themselves are powerful ones).

On top of that, this appears to be inconsistent with the previous implementation, which delegated the decision to `AbilityPointCosts` if the soldier had a Brigadier rank.

I agree that things are a bit muddy here. I would prefer to see the default behaviour match the base game's AP awards, although now that I think about it, that's specific to faction soldiers. Anyway, I'm willing to defer to folks with more experience with the vanilla game, since LWOTC overrides most of this behaviour anyway.

Reply:
Well, since base game code itself does a hardcoded rank check `if (SoldierRank >= 7)`, I think we can do the same here. I'll adjust.